### ISSUE TITLE
Missed optimization of Loop Unswitch (could be a regression)

### ISSUE THREAD

**Issue by ZY546** [llvm/llvm-project#67813]
Hello, we found some optimizations (regarding Loop Unswitch) that LLVM may have missed. We would greatly appreicate if you can take a look and let us know you think.

The control flow graph for this code is simple. Additionally, these optimizations work as expected on llvm-10.0.0.

Here are two examples.

Example 1:
https://godbolt.org/z/GKr3EP5WE

Given the following code: 
```c++
void test(int var_0, int var_1, int var_2, int var_3) {
    for (int i_0 = var_7 - 2121007231; i_0 < var_3 + 1860531184; i_0 += -838747044 - var_3 - var_0 + 543053277) 
    {
        var_8 += -1112137106 - var_3 + 5 - 333186413;
        var_1 += var_1;
        var_10 += var_2 + var_0;
        if (var_3)
        {
            var_20 += var_0 + var_1 + 16777216 - 23;
            var_28 += var_0 + var_8 - var_3 + var_0 - var_0;
            var_2 += var_16 + var_10 + var_30 - var_0 + var_1;
        }

        var_34 += var_2 - var_1 / var_3 - var_1 + var_2 - var_2;
        var_35 += var_34 & var_3 | ((var_1 + 2147483647) >> (var_0 + 545253533)) + var_12;
        var_0 += var_1 + 1694456005 + var_3 + var_0 + var_20 + var_1 - var_3;
        var_38 += var_2 - 2147483643 + 2147483647 + (var_1 && var_0) + 1369228270 + var_2;
        var_39 += var_2 + var_35 - var_3 + var_26;
        var_41 += var_19 + var_0 + var_28;
    }
}
```

Because `var_3` is a loop invariant so we can hoist the if condition out of the loop, but clang-trunk -O3 dose not (main parts):
```asm
.LBB0_2:                                # %for.body
        ...   
        jne     .LBB0_6                              #if-else
        ...
```

Example 2:
https://godbolt.org/z/q8GaPfvPr

Given the following code: 
```c++
void test(int var_0, int var_1, int var_2, int var_3, int var_4) {
    for (int i_0 = 0; i_0 < 4 + -1986643035 + 1 - 1 + 1986643054; i_0 += var_3 - var_1 - var_3 - var_10 + var_10 - 636989219 + var_2 - 691490232 - 1995983094) {
        var_11 += 691490244 + 1118573265 - 1740428269 - 15 + var_1 - var_1 + var_3 - var_1;
        var_2 += 691490244 - var_2 + var_9 - var_4 + 524287 + var_0 - 130944 - var_8;
        var_2 += var_1 - var_2 - 1278891492 + var_0 + var_10 / var_2 - 1060987095 / var_10;
        
        if (var_4) {
            var_15 += var_3 + var_2 + var_12 - var_10 + var_4 - var_10 + var_2 + var_0;
        }
        
        var_0 += var_2 + var_2 + var_17 - var_11 + 908521917 + 691490232 + var_18 + var_2;
        var_3 += -2147483627 - var_15 + var_4 - -4 - 1026228299;
        var_25 += var_3 / var_0 / (var_10 / var_9) * (28 / var_0 / (var_8 / var_3));
    }
}
```

Similarly, `var_4` is an loop invariant and should be hoisted out of the loop as well.
But clang-trunk -O3 does not:
```asm
.LBB0_1:                                # %for.body
        ...
        je      .LBB0_3                                #if-else
        ...
```

Thank you very much for your time and effort! We look forward to hearing from you.

**Comment by nikic** [comment]
Output from `-mllvm -debug-only=simple-loop-unswitch`:
```
Unswitching loop in _Z4testiiiii: Loop at depth 1 containing: %for.body<header>,%if.then,%if.end<latch><exiting>

  Trying to unswitch branch:   br i1 %tobool.not, label %if.end, label %if.then, !dbg !43
   Branch doesn't exit the loop!
Considering 1 non-trivial loop invariant conditions for unswitching.
  Total loop cost: 59
  Computed multiplier  1 (siblings 1 * clones 1) for unswitch candidate:   br i1 %tobool.not, label %if.end, label %if.then, !dbg !43
  Computed cost of 50 (multiplier: 1) for unswitch candidate:   br i1 %tobool.not, label %if.end, label %if.then, !dbg !43
Cannot unswitch, lowest cost found: 50
```
So just a cost issue?

**Comment by ZY546** [comment]
> Output from `-mllvm -debug-only=simple-loop-unswitch`:
> 
> ```
> Unswitching loop in _Z4testiiiii: Loop at depth 1 containing: %for.body<header>,%if.then,%if.end<latch><exiting>
> 
>   Trying to unswitch branch:   br i1 %tobool.not, label %if.end, label %if.then, !dbg !43
>    Branch doesn't exit the loop!
> Considering 1 non-trivial loop invariant conditions for unswitching.
>   Total loop cost: 59
>   Computed multiplier  1 (siblings 1 * clones 1) for unswitch candidate:   br i1 %tobool.not, label %if.end, label %if.then, !dbg !43
>   Computed cost of 50 (multiplier: 1) for unswitch candidate:   br i1 %tobool.not, label %if.end, label %if.then, !dbg !43
> Cannot unswitch, lowest cost found: 50
> ```
> 
> So just a cost issue?

Hi nikic, thank you for your answer, but there seems to be a problem here.

We compared the llvm-10.0.0 source code with the llvm-trunk source code and found the same code here:

```C
static cl::opt<int>
    UnswitchThreshold("unswitch-threshold", cl::init(50), cl::Hidden,
                      cl::desc("The cost threshold for unswitching a loop."));
```

That is, the code threshold is the same. But for the same code, llvm-10.0.0 does a loop Unswitch.

So there seems to be something else going on here. 

For example, in [Example 1](https://godbolt.org/z/GKr3EP5WE), line 34:
`var_0 += var_1 + 1694456005 + var_3 + var_0 + var_20 + var_1 - var_3;`

Where `var_3 - var_3` is 0, so the evaluation of this expression should eliminate var_3.

llvm-10.0.0 does a good job of this, but llvm-trunk does not.

So it seems that this is not just a cost issue, but an unnecessary cost added by some missed optimizations? This may be a regression problem as well.

(If we manually [eliminate var_3-var_3](https://godbolt.org/z/3Tj89abro), then it will work as expected.)

**Comment by nikic** [comment]
> Hi @nikic , thank you for your answer, but there seems to be a problem here.
> 
> We compared the llvm-10.0.0 source code with the llvm-trunk source code and found the same code here:
> 
> ```c
> static cl::opt<int>
>     UnswitchThreshold("unswitch-threshold", cl::init(50), cl::Hidden,
>                       cl::desc("The cost threshold for unswitching a loop."));
> ```
> 
> That is, the code threshold is the same. But for the same code, llvm-10.0.0 does a loop Unswitch.
> 
> So there seems to be something else going on here.

That's not really unexpected, given that the cost here sits directly at the edge, so any minor perturbation in cost modelling will move it one way or the other. Unless there is reason to believe that one choice is more correct than the other here (which I don't think there is, this is exactly the kind of case where loop unswitching profitability is unclear), there is nothing actionable here.

> For example, in [Example 1](https://godbolt.org/z/GKr3EP5WE), line 34: `var_0 += var_1 + 1694456005 + var_3 + var_0 + var_20 + var_1 - var_3;`
> 
> Where `var_3 - var_3` is 0, so the evaluation of this expression should eliminate var_3.
> 
> llvm-10.0.0 does a good job of this, but llvm-trunk does not.
> 
> So it seems that this is not just a cost issue, but an unnecessary cost added by some missed optimizations? This may be a regression problem as well.
> 
> (If we manually [eliminate var_3-var_3](https://godbolt.org/z/3Tj89abro), then it will work as expected.)

This is a missed optimization, but it has no relation to loop unswitching.

-----

Generally, I would ask you to stop reporting these fuzzer generated "missed optimization" issues, unless you are willing to put in more effort into providing a high quality issue report. A missed optimization report should have a clean test case (no weird constants, no weird variable names, no weird expression trees) and a root cause analysis -- at least sufficient root cause analysis to clearly distinguish one issue from another.

For example, you have opened a whole slew of LICM issues that basically boil down to "LICM is not doing something it could do, I have no idea why, I have no idea whether this is the same issue as one of the ten others I reported or not". If the report is instead "LICM fails to hoist expression that requires integer add reassociation", with a minimal example of that pattern, that becomes an issue report someone may actually look at and fix.

You have opened a lot of issues, and I can honestly say that I did not find a single one of them useful. I don't think any of these issue reports have resulted in someone implementing a new optimization in LLVM. If you want that to change, the nature of these reports needs to change.
